crates/core/tedge/src/cli/certificate/cli.rs

```diff
@@ -193,16 +191,247 @@ pub enum UploadCertCli {
     },
 }
 
-fn get_device_id_from_config(
+/// Returns the device ID from the config if no ID is provided by CLI
+fn get_device_id(
+    id: Option<String>,
     config: &TEdgeConfig,
+    config_location: &TEdgeConfigLocation,
     cloud: &Option<Cloud>,
-) -> Result<String, ReadError> {
-    let id = match cloud {
-        None => config.device.id(),
-        Some(Cloud::C8y(profile)) => config.c8y.try_get(profile.as_deref())?.device.id(),
-        Some(Cloud::Azure(profile)) => config.az.try_get(profile.as_deref())?.device.id(),
-        Some(Cloud::Aws(profile)) => config.aws.try_get(profile.as_deref())?.device.id(),
-    }?
-    .to_owned();
-    Ok(id)
+) -> Result<String, anyhow::Error> {
+    match (id, config.device_id(cloud.as_ref()).ok()) {
+        (None, None) => Err(anyhow!(
+            "No device ID is provided. Use `--device-id <name>` option to specify the device ID."
+        )),
+        (None, Some(config_id)) => Ok(config_id.into()),
+        (Some(input_id), None) => Ok(input_id),
+        (Some(input_id), Some(config_id)) if input_id == config_id => Ok(input_id),
+        (Some(input_id), Some(_config_id)) => {
+            // if ids from input and TEdgeConfigReader mismatch, need to dive into TEdgeConfigDto level
+            validate_input_id_with_dto(&input_id, config_location, cloud)
+        }
```

Comment:
I don't really get the changes introduced by this commit https://github.com/thin-edge/thin-edge.io/pull/3326/commits/09bf5f92e1dfef2aaee9e44ed90b497b584f960b.

- First, this doesn't fix the issue with `tedge cert create c8y --profile foo` requesting a device id while `tedge config get device.id` returns one (this issue arises when the device id is read from the main device certificate and not the config)
- Second, it seems that there is no difference in behavior between the two ways to read the config (`config.device_id(cloud)` vs what is done inside `validate_input_id_with_dto`). (I have to double check that)